File: tallow-cli/internal/network/stun.go
package network

import (
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"net"
	"time"
)

// STUN message types
const (
	stunMsgTypeBindingRequest  = 0x0001
	stunMsgTypeBindingResponse = 0x0101
	stunMsgTypeBindingError    = 0x0111
)

// STUN attribute types
const (
	stunAttrMappedAddress    = 0x0001
	stunAttrXORMappedAddress = 0x0020
)

// STUN magic cookie
const stunMagicCookie = 0x2112A442

// Default STUN servers
var DefaultSTUNServers = []string{
	"stun.l.google.com:19302",
	"stun1.l.google.com:19302",
	"stun2.l.google.com:19302",
	"stun.cloudflare.com:3478",
}

// NATType represents the type of NAT
type NATType int

const (
	NATTypeUnknown NATType = iota
	NATTypeNone            // No NAT (public IP)
	NATTypeFullCone
	NATTypeRestrictedCone
	NATTypePortRestricted
	NATTypeSymmetric
)

func (n NATType) String() string {
	switch n {
	case NATTypeNone:
		return "No NAT (Public IP)"
	case NATTypeFullCone:
		return "Full Cone NAT"
	case NATTypeRestrictedCone:
		return "Restricted Cone NAT"
	case NATTypePortRestricted:
		return "Port Restricted Cone NAT"
	case NATTypeSymmetric:
		return "Symmetric NAT"
	default:
		return "Unknown"
	}
}

// STUNResult contains the result of a STUN query
type STUNResult struct {
	PublicIP   string
	PublicPort int
	NATType    NATType
	LocalIP    string
	LocalPort  int
}

// STUNClient performs STUN queries
type STUNClient struct {
	servers []string
	timeout time.Duration
}

// NewSTUNClient creates a new STUN client
func NewSTUNClient(servers []string) *STUNClient {
	if len(servers) == 0 {
		servers = DefaultSTUNServers
	}
	return &STUNClient{
		servers: servers,
		timeout: 5 * time.Second,
	}
}

// GetPublicAddress queries STUN servers to get the public address
func (c *STUNClient) GetPublicAddress(ctx context.Context) (*STUNResult, error) {
	// Create UDP socket
	conn, err := net.ListenUDP("udp4", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create UDP socket: %w", err)
	}
	defer conn.Close()

	localAddr := conn.LocalAddr().(*net.UDPAddr)

	// Try each STUN server
	for _, server := range c.servers {
		result, err := c.queryServer(ctx, conn, server)
		if err == nil {
			result.LocalIP = localAddr.IP.String()
			result.LocalPort = localAddr.Port
			return result, nil
		}
	}

	return nil, errors.New("all STUN servers failed")
}

// queryServer queries a single STUN server
func (c *STUNClient) queryServer(ctx context.Context, conn *net.UDPConn, server string) (*STUNResult, error) {
	serverAddr, err := net.ResolveUDPAddr("udp4", server)
	if err != nil {
		return nil, err
	}

	// Build STUN binding request
	request := c.buildBindingRequest()

	// Set timeout
	deadline, ok := ctx.Deadline()
	if !ok {
		deadline = time.Now().Add(c.timeout)
	}
	conn.SetDeadline(deadline)

	// Send request
	if _, err := conn.WriteToUDP(request, serverAddr); err != nil {
		return nil, err
	}

	// Read response
	response := make([]byte, 1024)
	n, _, err := conn.ReadFromUDP(response)
	if err != nil {
		return nil, err
	}

	// Parse response
	return c.parseBindingResponse(response[:n])
}

// buildBindingRequest builds a STUN binding request
func (c *STUNClient) buildBindingRequest() []byte {
	// 20 bytes header
	request := make([]byte, 20)

	// Message type: Binding Request
	binary.BigEndian.PutUint16(request[0:2], stunMsgTypeBindingRequest)

	// Message length: 0 (no attributes)
	binary.BigEndian.PutUint16(request[2:4], 0)

	// Magic cookie
	binary.BigEndian.PutUint32(request[4:8], stunMagicCookie)

	// Transaction ID (96 bits)
	// Use current time for simplicity (not cryptographically random but sufficient for STUN)
	now := time.Now().UnixNano()
	binary.BigEndian.PutUint64(request[8:16], uint64(now))
	binary.BigEndian.PutUint32(request[16:20], uint32(now>>32))

	return request
}

// parseBindingResponse parses a STUN binding response
func (c *STUNClient) parseBindingResponse(data []byte) (*STUNResult, error) {
	if len(data) < 20 {
		return nil, errors.New("response too short")
	}

	// Check message type
	msgType := binary.BigEndian.Uint16(data[0:2])
	if msgType != stunMsgTypeBindingResponse {
		if msgType == stunMsgTypeBindingError {
			return nil, errors.New("STUN binding error")
		}
		return nil, errors.New("unexpected response type")
	}

	// Check magic cookie
	cookie := binary.BigEndian.Uint32(data[4:8])
	if cookie != stunMagicCookie {
		return nil, errors.New("invalid magic cookie")
	}

	// Parse attributes
	msgLen := binary.BigEndian.Uint16(data[2:4])
	if int(msgLen)+20 > len(data) {
		return nil, errors.New("invalid message length")
	}

	offset := 20
	for offset < 20+int(msgLen) {
		if offset+4 > len(data) {
			break
		}

		attrType := binary.BigEndian.Uint16(data[offset : offset+2])
		attrLen := binary.BigEndian.Uint16(data[offset+2 : offset+4])
		offset += 4

		if offset+int(attrLen) > len(data) {
			break
		}

		attrData := data[offset : offset+int(attrLen)]

		if attrType == stunAttrXORMappedAddress {
			return c.parseXORMappedAddress(attrData, data[4:8])
		} else if attrType == stunAttrMappedAddress {
			return c.parseMappedAddress(attrData)
		}

		// Align to 4 bytes
		offset += int(attrLen)
		if attrLen%4 != 0 {
			offset += 4 - int(attrLen%4)
		}
	}

	return nil, errors.New("no mapped address in response")
}

// parseXORMappedAddress parses an XOR-MAPPED-ADDRESS attribute
func (c *STUNClient) parseXORMappedAddress(data, transactionID []byte) (*STUNResult, error) {
	if len(data) < 8 {
		return nil, errors.New("invalid XOR-MAPPED-ADDRESS")
	}

	family := data[1]
	port := binary.BigEndian.Uint16(data[2:4]) ^ uint16(stunMagicCookie>>16)

	if family == 0x01 { // IPv4
		ip := make([]byte, 4)
		magicBytes := make([]byte, 4)
		binary.BigEndian.PutUint32(magicBytes, stunMagicCookie)
		for i := 0; i < 4; i++ {
			ip[i] = data[4+i] ^ magicBytes[i]
		}
		return &STUNResult{
			PublicIP:   net.IP(ip).String(),
			PublicPort: int(port),
		}, nil
	}

	return nil, errors.New("unsupported address family")
}

// parseMappedAddress parses a MAPPED-ADDRESS attribute
func (c *STUNClient) parseMappedAddress(data []byte) (*STUNResult, error) {
	if len(data) < 8 {
		return nil, errors.New("invalid MAPPED-ADDRESS")
	}

	family := data[1]
	port := binary.BigEndian.Uint16(data[2:4])

	if family == 0x01 { // IPv4
		ip := net.IP(data[4:8])
		return &STUNResult{
			PublicIP:   ip.String(),
			PublicPort: int(port),
		}, nil
	}

	return nil, errors.New("unsupported address family")
}

// DetectNATType performs NAT type detection (simplified)
func (c *STUNClient) DetectNATType(ctx context.Context) (*STUNResult, error) {
	result, err := c.GetPublicAddress(ctx)
	if err != nil {
		return nil, err
	}

	// Compare local and public IP
	localIP, err := GetLocalIP()
	if err != nil {
		result.NATType = NATTypeUnknown
		return result, nil
	}

	if localIP == result.PublicIP {
		result.NATType = NATTypeNone
	} else {
		// Simplified: assume port-restricted for now
		// Full NAT detection requires multiple STUN queries
		result.NATType = NATTypePortRestricted
	}

	return result, nil
}

File: crates/tallow-net/src/nat/stun.rs
//! STUN client for NAT type discovery
//!
//! Implements a minimal STUN Binding Request (RFC 5389) to discover
//! the public IP address and port mapping of the client.

use crate::error::NetworkError;
use crate::Result;
use std::net::SocketAddr;
use tokio::net::UdpSocket;
use tokio::time::{timeout, Duration};

/// Well-known STUN servers
pub const GOOGLE_STUN: &str = "stun.l.google.com:19302";
pub const CLOUDFLARE_STUN: &str = "stun.cloudflare.com:3478";

/// STUN magic cookie (RFC 5389)
const MAGIC_COOKIE: u32 = 0x2112A442;

/// STUN message types
const BINDING_REQUEST: u16 = 0x0001;
const BINDING_RESPONSE: u16 = 0x0101;

/// STUN attribute types
const ATTR_MAPPED_ADDRESS: u16 = 0x0001;
const ATTR_XOR_MAPPED_ADDRESS: u16 = 0x0020;

/// STUN client for discovering public address
#[derive(Debug)]
pub struct StunClient {
    server: SocketAddr,
}

/// STUN binding result
#[derive(Debug, Clone)]
pub struct StunResult {
    /// Public (mapped) address as seen by the STUN server
    pub mapped_addr: SocketAddr,
    /// Local address used
    pub local_addr: SocketAddr,
}

impl StunClient {
    /// Create a new STUN client targeting a specific server
    pub fn new(server: SocketAddr) -> Self {
        Self { server }
    }

    /// Create a STUN client from a hostname (resolves DNS)
    pub async fn from_hostname(hostname: &str) -> Result<Self> {
        let addr = tokio::net::lookup_host(hostname)
            .await
            .map_err(|e| {
                NetworkError::DnsResolution(format!(
                    "DNS resolution failed for {}: {}",
                    hostname, e
                ))
            })?
            .next()
            .ok_or_else(|| {
                NetworkError::DnsResolution(format!("No addresses found for {}", hostname))
            })?;

        Ok(Self { server: addr })
    }

    /// Discover public address via STUN Binding Request
    pub async fn discover_public_address(&self) -> Result<StunResult> {
        let socket = UdpSocket::bind("0.0.0.0:0")
            .await
            .map_err(|e| {
                NetworkError::NatTraversal(format!("Failed to bind UDP socket: {}", e))
            })?;

        let local_addr = socket.local_addr().map_err(|e| {
            NetworkError::NatTraversal(format!("Failed to get local address: {}", e))
        })?;

        // Build STUN Binding Request
        let transaction_id: [u8; 12] = rand::random();
        let request = build_binding_request(&transaction_id);

        // Send request
        socket
            .send_to(&request, self.server)
            .await
            .map_err(|e| {
                NetworkError::NatTraversal(format!("Failed to send STUN request: {}", e))
            })?;

        // Wait for response (3 second timeout)
        let mut buf = [0u8; 576];
        let (len, _) = timeout(Duration::from_secs(3), socket.recv_from(&mut buf))
            .await
            .map_err(|_| NetworkError::Timeout)?
            .map_err(|e| {
                NetworkError::NatTraversal(format!("Failed to receive STUN response: {}", e))
            })?;

        // Parse response
        let mapped_addr = parse_binding_response(&buf[..len], &transaction_id)?;

        Ok(StunResult {
            mapped_addr,
            local_addr,
        })
    }
}

/// Build a STUN Binding Request message
fn build_binding_request(transaction_id: &[u8; 12]) -> Vec<u8> {
    let mut msg = Vec::with_capacity(20);

    // Message Type: Binding Request (0x0001)
    msg.extend_from_slice(&BINDING_REQUEST.to_be_bytes());
    // Message Length: 0 (no attributes)
    msg.extend_from_slice(&0u16.to_be_bytes());
    // Magic Cookie
    msg.extend_from_slice(&MAGIC_COOKIE.to_be_bytes());
    // Transaction ID (12 bytes)
    msg.extend_from_slice(transaction_id);

    msg
}

/// Parse a STUN Binding Response
fn parse_binding_response(data: &[u8], expected_txn: &[u8; 12]) -> Result<SocketAddr> {
    if data.len() < 20 {
        return Err(NetworkError::NatTraversal(
            "STUN response too short".to_string(),
        ));
    }

    // Check message type
    let msg_type = u16::from_be_bytes([data[0], data[1]]);
    if msg_type != BINDING_RESPONSE {
        return Err(NetworkError::NatTraversal(format!(
            "Unexpected STUN message type: 0x{:04x}",
            msg_type
        )));
    }

    let msg_len = u16::from_be_bytes([data[2], data[3]]) as usize;

    // Check magic cookie
    let cookie = u32::from_be_bytes([data[4], data[5], data[6], data[7]]);
    if cookie != MAGIC_COOKIE {
        return Err(NetworkError::NatTraversal(
            "Invalid STUN magic cookie".to_string(),
        ));
    }

    // Check transaction ID
    if data[8..20] != *expected_txn {
        return Err(NetworkError::NatTraversal(
            "STUN transaction ID mismatch".to_string(),
        ));
    }

    // Parse attributes
    let attrs = &data[20..20 + msg_len.min(data.len() - 20)];
    let mut i = 0;

    while i + 4 <= attrs.len() {
        let attr_type = u16::from_be_bytes([attrs[i], attrs[i + 1]]);
        let attr_len = u16::from_be_bytes([attrs[i + 2], attrs[i + 3]]) as usize;
        let attr_data = &attrs[i + 4..];

        if attr_data.len() < attr_len {
            break;
        }

        match attr_type {
            ATTR_XOR_MAPPED_ADDRESS => {
                return parse_xor_mapped_address(&attr_data[..attr_len], data);
            }
            ATTR_MAPPED_ADDRESS => {
                return parse_mapped_address(&attr_data[..attr_len]);
            }
            _ => {}
        }

        // Attributes are padded to 4-byte boundaries
        let padded_len = (attr_len + 3) & !3;
        i += 4 + padded_len;
    }

    Err(NetworkError::NatTraversal(
        "No mapped address in STUN response".to_string(),
    ))
}

/// Parse XOR-MAPPED-ADDRESS attribute
fn parse_xor_mapped_address(data: &[u8], header: &[u8]) -> Result<SocketAddr> {
    if data.len() < 8 {
        return Err(NetworkError::NatTraversal(
            "XOR-MAPPED-ADDRESS too short".to_string(),
        ));
    }

    let family = data[1];
    let xport = u16::from_be_bytes([data[2], data[3]]) ^ (MAGIC_COOKIE >> 16) as u16;

    match family {
        0x01 => {
            // IPv4
            let xaddr = u32::from_be_bytes([data[4], data[5], data[6], data[7]]) ^ MAGIC_COOKIE;
            let ip = std::net::Ipv4Addr::from(xaddr);
            Ok(SocketAddr::new(ip.into(), xport))
        }
        0x02 => {
            // IPv6
            if data.len() < 20 {
                return Err(NetworkError::NatTraversal(
                    "XOR-MAPPED-ADDRESS IPv6 too short".to_string(),
                ));
            }
            let mut xaddr = [0u8; 16];
            xaddr.copy_from_slice(&data[4..20]);
            // XOR with magic cookie + transaction ID
            for j in 0..4 {
                xaddr[j] ^= header[4 + j];
            }
            for j in 0..12 {
                xaddr[4 + j] ^= header[8 + j];
            }
            let ip = std::net::Ipv6Addr::from(xaddr);
            Ok(SocketAddr::new(ip.into(), xport))
        }
        _ => Err(NetworkError::NatTraversal(format!(
            "Unknown address family: {}",
            family
        ))),
    }
}

/// Parse MAPPED-ADDRESS attribute
fn parse_mapped_address(data: &[u8]) -> Result<SocketAddr> {
    if data.len() < 8 {
        return Err(NetworkError::NatTraversal(
            "MAPPED-ADDRESS too short".to_string(),
        ));
    }

    let family = data[1];
    let port = u16::from_be_bytes([data[2], data[3]]);

    match family {
        0x01 => {
            let ip = std::net::Ipv4Addr::new(data[4], data[5], data[6], data[7]);
            Ok(SocketAddr::new(ip.into(), port))
        }
        0x02 => {
            if data.len() < 20 {
                return Err(NetworkError::NatTraversal(
                    "MAPPED-ADDRESS IPv6 too short".to_string(),
                ));
            }
            let mut addr = [0u8; 16];
            addr.copy_from_slice(&data[4..20]);
            let ip = std::net::Ipv6Addr::from(addr);
            Ok(SocketAddr::new(ip.into(), port))
        }
        _ => Err(NetworkError::NatTraversal(format!(
            "Unknown address family: {}",
            family
        ))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_binding_request() {
        let txn = [1u8; 12];
        let msg = build_binding_request(&txn);
        assert_eq!(msg.len(), 20);
        assert_eq!(msg[0], 0x00);
        assert_eq!(msg[1], 0x01);
        assert_eq!(msg[2], 0x00);
        assert_eq!(msg[3], 0x00);
        assert_eq!(
            u32::from_be_bytes([msg[4], msg[5], msg[6], msg[7]]),
            MAGIC_COOKIE
        );
        assert_eq!(&msg[8..20], &txn);
    }

    #[test]
    fn test_parse_xor_mapped_ipv4() {
        let txn = [0xAA; 12];
        let mut resp = Vec::new();
        resp.extend_from_slice(&BINDING_RESPONSE.to_be_bytes());
        resp.extend_from_slice(&12u16.to_be_bytes());
        resp.extend_from_slice(&MAGIC_COOKIE.to_be_bytes());
        resp.extend_from_slice(&txn);
        // XOR-MAPPED-ADDRESS attribute
        resp.extend_from_slice(&ATTR_XOR_MAPPED_ADDRESS.to_be_bytes());
        resp.extend_from_slice(&8u16.to_be_bytes());
        resp.push(0x00); // reserved
        resp.push(0x01); // IPv4
        let xport = 8080u16 ^ 0x2112;
        resp.extend_from_slice(&xport.to_be_bytes());
        let ip_raw = u32::from_be_bytes([192, 168, 1, 1]) ^ MAGIC_COOKIE;
        resp.extend_from_slice(&ip_raw.to_be_bytes());

        let result = parse_binding_response(&resp, &txn).unwrap();
        assert_eq!(
            result.ip(),
            std::net::IpAddr::V4(std::net::Ipv4Addr::new(192, 168, 1, 1))
        );
        assert_eq!(result.port(), 8080);
    }
}
